cmd/utils/history_test.go

```diff
@@ -0,0 +1,200 @@
+		t.Fatalf("error insterting chain: %v", err)
+	}
+
+	// Make temp directory for era files.
+	dir, err := os.MkdirTemp("", "history-export-test")
+	if err != nil {
+		t.Fatalf("error creating temp test directory: %v", err)
+	}
+	defer os.RemoveAll(dir)
+
+	// Export history to temp directory.
+	if err := ExportHistory(chain, dir, 0, count, step); err != nil {
+		t.Fatalf("error exporting history: %v", err)
+	}
+
+	// Read checksums.
+	b, err := os.ReadFile(path.Join(dir, "checksums.txt"))
+	if err != nil {
+		t.Fatalf("failed to read checksums: %v", err)
+	}
+	checksums := strings.Split(string(b), "\n")
+
+	// Verify each Era.
+	entries, _ := era.ReadDir(dir, "mainnet")
+	for i, filename := range entries {
+		b, err := os.ReadFile(path.Join(dir, filename))
+		if err != nil {
+			t.Fatalf("error opening era file: %v", err)
+		}
+		if want, got := common.HexToHash(checksums[i]), common.Hash(sha256.Sum256(b)); want != got {
+			t.Fatalf("checksum %d does not match: got %s, want %s", i, got, want)
+		}
+		r := era.NewReader(bytes.NewReader(b))
+		for j := 0; ; j += 1 {
+			b, r, err := r.Read()
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				t.Fatalf("error reading era file %d: %v", i, err)
+			}
+			var (
+				num  = i*int(step) + j
+				want = chain.GetBlockByNumber(uint64(num))
+			)
+			if want, got := uint64(num), b.NumberU64(); want != got {
+				t.Fatalf("blocks out of order: want %d, got %d", want, got)
+			}
+			if want.Hash() != b.Hash() {
+				t.Fatalf("block hash mistmatch %d: want %s, got %s", i+j, want.Hash().Hex(), b.Hash().Hex())
+			}
+			if got := types.DeriveSha(b.Transactions(), trie.NewStackTrie(nil)); got != want.TxHash() {
+				t.Fatalf("tx hash %d mismatch: want %s, got %s", i+j, want.TxHash(), got)
+			}
+			if got := types.CalcUncleHash(b.Uncles()); got != want.UncleHash() {
+				t.Fatalf("uncle hash %d mismatch: want %s, got %s", i+j, want.UncleHash(), got)
+			}
+			if got := types.DeriveSha(r, trie.NewStackTrie(nil)); got != want.ReceiptHash() {
+				t.Fatalf("receipt root %d mismatch: want %s, got %s", i+j, want.ReceiptHash(), got)
+			}
+		}
+	}
+
+	// Now import Era.
+	freezer := t.TempDir()
+	db2, err := rawdb.NewDatabaseWithFreezer(rawdb.NewMemoryDatabase(), freezer, "", false)
+	if err != nil {
+		panic(err)
+	}
+	t.Cleanup(func() {
+		db2.Close()
+	})
+
+	genesis.MustCommit(db2)
+	imported, err := core.NewBlockChain(db2, nil, genesis, nil, ethash.NewFaker(), vm.Config{}, nil, nil)
+	if err != nil {
+		t.Fatalf("unable to initialize chain: %v", err)
+	}
+	if err := ImportHistory(imported, db2, dir, "mainnet"); err != nil {
+		t.Fatalf("failed to import chain: %v", err)
+	}
+	if have, want := imported.CurrentHeader(), chain.CurrentHeader(); have.Hash() != want.Hash() {
+		t.Fatalf("imported chain does not match expected, have (%d, %s) want (%d, %s)", have.Number, have.Hash(), want.Number, want.Hash())
+	}
+}
+
+func BenchmarkHistoryImport(b *testing.B) {
```

Comment:
I get the intention of this benchmark, benchmark the import of a certain export you have locally. I think benchmarks should be self contained, so you don't need some extra piece of data.

Reply:
WDYT about something like this:

```

func BenchmarkImportAndExport(b *testing.B) {
	var (
		key, _  = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
		address = crypto.PubkeyToAddress(key.PublicKey)
		genesis = &core.Genesis{
			Config: params.TestChainConfig,
			Alloc:  core.GenesisAlloc{address: {Balance: big.NewInt(1000000000000000000)}},
		}
		signer = types.LatestSigner(genesis.Config)
	)

	// Generate chain.
	db, blocks, _ := core.GenerateChainWithGenesis(genesis, ethash.NewFaker(), int(b.N), func(i int, g *core.BlockGen) {
		if i == 0 {
			return
		}
		tx, err := types.SignNewTx(key, signer, &types.DynamicFeeTx{
			ChainID:    genesis.Config.ChainID,
			Nonce:      uint64(i - 1),
			GasTipCap:  common.Big0,
			GasFeeCap:  g.PrevBlock(0).BaseFee(),
			Gas:        50000,
			To:         &common.Address{0xaa},
			Value:      big.NewInt(int64(i)),
			Data:       nil,
			AccessList: nil,
		})
		if err != nil {
			b.Fatalf("error creating tx: %v", err)
		}
		g.AddTx(tx)
	})

	// Initialize BlockChain.
	chain, err := core.NewBlockChain(db, nil, genesis, nil, ethash.NewFaker(), vm.Config{}, nil, nil)
	if err != nil {
		b.Fatalf("unable to initialize chain: %v", err)
	}
	if _, err := chain.InsertChain(blocks); err != nil {
		b.Fatalf("error insterting chain: %v", err)
	}

	// Make temp directory for era files.
	dir, err := os.MkdirTemp("", "history-export-test")
	if err != nil {
		b.Fatalf("error creating temp test directory: %v", err)
	}
	defer os.RemoveAll(dir)

	b.ResetTimer()
	// Export history to temp directory.
	if err := ExportHistory(chain, dir, 0, uint64(b.N), step); err != nil {
		b.Fatalf("error exporting history: %v", err)
	}
	b.StopTimer()

	// Now import Era.
	freezer := b.TempDir()
	db2, err := rawdb.NewDatabaseWithFreezer(rawdb.NewMemoryDatabase(), freezer, "", false)
	if err != nil {
		panic(err)
	}
	b.Cleanup(func() {
		db2.Close()
	})

	genesis.MustCommit(db2)
	imported, err := core.NewBlockChain(db2, nil, genesis, nil, ethash.NewFaker(), vm.Config{}, nil, nil)
	if err != nil {
		b.Fatalf("unable to initialize chain: %v", err)
	}
	b.StartTimer()
	if err := ImportHistory(imported, db2, dir, "mainnet"); err != nil {
		b.Fatalf("failed to import chain: %v", err)
	}
	if have, want := imported.CurrentHeader(), chain.CurrentHeader(); have.Hash() != want.Hash() {
		b.Fatalf("imported chain does not match expected, have (%d, %s) want (%d, %s)", have.Number, have.Hash(), want.Number, want.Hash())
	}
}
```

Reply:
I think I can just delete this - I had a benchmark similar to this before, but changed it to run the bench on some actual mainnet blocks to get a more realistic bench.